### DISCUSSION TITLE
Intermittently slow ListObjects performance

### DISCUSSION THREAD

**Discussion by Jazz9** [minio/minio#19320] [Q&A] [answered]
I've been looking into a intermittent slow listobjects issue on our storage. Sometimes it lists a 1000 object page in < 100ms and other times takes 30 + seconds. Most of the time it takes < 1s.

I've tried a dec/2023 and latest builds without change. The minio is configured as a single node with a single enterprise nvme and the iops / queue depth / cpu / memory is not the bottleneck.

The trace is showing that the storage.walkdir sometimes takes too long and the client get a 499 as the context deadline expires.

The storage has 500+ million small files and I wondered if the GO walkdir was the issue. I wrote a test GO 1.18.1 app to walk the filesystem and it performs very well. It will still take time to walk the complete tree at about 200,000 - 300,000 objects a second.

The maximum files in a folder is 1.6M and the test app walks in 11 seconds.

It 'feels like' a cache locking or similar issue. Is there any further debugging that is suggested to try improve the listobjects performance?

**Comment by klauspost** [comment, marked as answer]
@Jazz9 Look into out [subscriptions](https://min.io/pricing) if you are looking for performance analysis. 

Alternatively you are welcome to look into the source code and submit a PR.

**Comment by Qubitium** [comment]
@Jazz9  Check `iostat` and see if you are seeing spikes in `iowait` value above `1.0` which would mean the nvme is backlogged in iop requests. Sometimes the nvme needs to pause the world to do self-cleanup/trim.
